### DISCUSSION TITLE
Proposal for Interpolation and Extrapolation Interfaces in System.Numerics

### DISCUSSION THREAD

**Discussion by StellarWolfEntertainment** [dotnet/csharplang#8463] [Language Ideas]
I apologize in advance if this post has landed in the wrong category.

I propose the introduction of generic interfaces for interpolation and extrapolation in the `System.Numerics` namespace. While the immediate use cases may seem limited, these interfaces can provide consistency and flexibility across numeric types.

Currently, handling interpolation and extrapolation in numeric operations lacks a standardized approach, which can lead to code duplication and inconsistencies.

The following interfaces could be implemented to streamline interpolation and extrapolation for numeric types:
```csharp
namespace System.Numerics;

public interface IInterpolable<TSelf, TReturn, TFactor> where TSelf : IInterpolable<TSelf, TReturn, TFactor>
{
    static abstract TReturn Interpolate(TSelf start, TSelf end, TFactor factor);
}

public interface IExtrapolable<TSelf, TReturn, TFactor> where TSelf : IExtrapolable<TSelf, TReturn, TFactor>
{
    static abstract TReturn Extrapolate(TSelf start, TSelf end, TFactor factor);
}

public interface IBlendable<TSelf, TReturn, TFactor> : IInterpolable<TSelf, TReturn, TFactor>, IExtrapolable<TSelf, TReturn, TFactor> where TSelf : IInterpolable<TSelf, TReturn, TFactor>, IExtrapolable<TSelf, TReturn, TFactor>, IBlendable<TSelf, TReturn, TFactor>
{
}
```

An example implementation for the Single type (I am going to only post the code relevant to the operations at hand): 
```csharp
public readonly struct Single : IBlendable<Single, Single, Single>
{
    // Interpolate method
    public static Single Interpolate(Single start, Single end, Single factor)
    {
        return Extrapolate(start, end, Single.Clamp(factor, 0.0f, 1.0f));
    }

    // Extrapolate method
    public static Single Extrapolate(Single start, Single end, Single factor)
    {
        return start + ((end - start) * factor);
    }
}
```
By adding these interfaces to the definition of numeric types in C#, we can provide a consistent approach to fairly common operations. Much like how the `ITrigonometricFunctions<T>` interface standardizes handling trigonometric operations with generic types, these interfaces will enhance the flexibility of generic math operations.

While the immediate utility might not be apparent, I believe this proposal will open the door for more robust and maintainable code in the future.

**Comment by colejohnson66** [comment]
API requests go in @dotnet/runtime. This repository is just for the language syntax and semantics.

**Reply by StellarWolfEntertainment** [reply]
In order to fix that, should I close this discussion and remake it there?

**Reply by colejohnson66** [reply]
Yep
